File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/RequestString.java
package spectable.common;

import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class RequestString {
    private static final String DNCString = "?DNC?";

    public String method;
    public String page;
    public String address;
    public String benchmark;
    public String format;

    public RequestString(String method, String page, String address, String benchmark, String format) {
        this.method = method;
        this.page = page;
        this.address = address;
        this.benchmark = benchmark;
        this.format = format;
    }

    /** Builds from the text form, e.g. Money as "25 USD". */
    public static RequestString fromText(String text) {
        java.util.List<String> parts = Tokens.require(text, 5, "Request");
        return new RequestString(parts.get(0), parts.get(1), parts.get(2), parts.get(3), parts.get(4));
    }

    public RequestString(String text) {
        RequestString parsed = fromText(text);
        this.method = parsed.method;
        this.page = parsed.page;
        this.address = parsed.address;
        this.benchmark = parsed.benchmark;
        this.format = parsed.format;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof RequestString)) return false;
        RequestString that = (RequestString) o;
        return (DNCString.equals(method) || DNCString.equals(that.method) || Objects.equals(method, that.method))
            && (DNCString.equals(page) || DNCString.equals(that.page) || Objects.equals(page, that.page))
            && (DNCString.equals(address) || DNCString.equals(that.address) || Objects.equals(address, that.address))
            && (DNCString.equals(benchmark) || DNCString.equals(that.benchmark) || Objects.equals(benchmark, that.benchmark))
            && (DNCString.equals(format) || DNCString.equals(that.format) || Objects.equals(format, that.format));
    }

    @Override
    public int hashCode() {
        return Objects.hash(method, page, address, benchmark, format);
    }

    @Override
    public String toString() {
        return Tokens.token(method) + " " + Tokens.token(page) + " " + Tokens.token(address) + " " + Tokens.token(benchmark) + " " + Tokens.token(format);
    }
}

File: CSharp/SpecStudioExamplesTestsCSharp/common/RequestString.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using production;

    public class RequestString
    {
        public string method;
        public string page;
        public string address;
        public string benchmark;
        public string format;

        public RequestString(string method, string page, string address, string benchmark, string format)
        {
            this.method = method;
            this.page = page;
            this.address = address;
            this.benchmark = benchmark;
            this.format = format;
        }

        /// <summary>Builds from the text form, e.g. Money as "25 USD".</summary>
        public static RequestString FromText(string text)
        {
            var parts = Tokens.Require(text, 5, "Request");
            return new RequestString(parts[0], parts[1], parts[2], parts[3], parts[4]);
        }

        public RequestString(string text)
        {
            var parsed = FromText(text);
            this.method = parsed.method;
            this.page = parsed.page;
            this.address = parsed.address;
            this.benchmark = parsed.benchmark;
            this.format = parsed.format;
        }

        public RequestTyped ToRequestTyped()
        {
            return new RequestTyped(
                this.method,
                this.page,
                this.address,
                this.benchmark,
                this.format
            );
        }

        public override string ToString()
        {
            return Tokens.Token(method) + " " + Tokens.Token(page) + " " + Tokens.Token(address) + " " + Tokens.Token(benchmark) + " " + Tokens.Token(format);
        }

        const string DNCString = "?DNC?";

        public override bool Equals(object? obj)
        {
            if (obj is not RequestString other) return false;
            return (DNCString == this.method || DNCString == other.method || object.Equals(this.method, other.method))
                && (DNCString == this.page || DNCString == other.page || object.Equals(this.page, other.page))
                && (DNCString == this.address || DNCString == other.address || object.Equals(this.address, other.address))
                && (DNCString == this.benchmark || DNCString == other.benchmark || object.Equals(this.benchmark, other.benchmark))
                && (DNCString == this.format || DNCString == other.format || object.Equals(this.format, other.format));
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.method);
            h.Add(this.page);
            h.Add(this.address);
            h.Add(this.benchmark);
            h.Add(this.format);
            return h.ToHashCode();
        }
    }
}
